katdal/categorical.py

```diff
@@ -16,124 +16,152 @@
-            items = sorted(set(ar))
-            return np.asarray(items)
-        else:
-            ar = np.asanyarray(ar).flatten()
+    def __str__(self):
+        """Longer human-friendly string representation of wrapped object."""
+        return str(self.unwrapped)
 
-    if ar.size == 0:
-        if return_inverse and return_index:
-            return ar, np.empty(0, np.bool), np.empty(0, np.bool)
-        elif return_inverse or return_index:
-            return ar, np.empty(0, np.bool)
+    def __eq__(self, other):
+        """Equality comparison operator."""
+        if isinstance(other, ComparableArrayWrapper):
+            other = other.unwrapped
+        if isinstance(self.unwrapped, np.ndarray) or isinstance(other, np.ndarray):
+            return np.array_equal(self.unwrapped, other)
         else:
-            return ar
+            return self.unwrapped == other
 
-    if return_inverse or return_index:
-        if return_index:
-            perm = ar.argsort(kind='mergesort')
-        else:
-            perm = ar.argsort()
-        aux = ar[perm]
-        flag = np.concatenate(([True], aux[1:] != aux[:-1]))
-        if return_inverse:
-            iflag = np.cumsum(flag) - 1
-            iperm = perm.argsort()
-            if return_index:
-                return aux[flag], perm[flag], iflag[iperm]
-            else:
-                return aux[flag], iflag[iperm]
-        else:
-            return aux[flag], perm[flag]
-    else:
-        ar.sort()
-        flag = np.concatenate(([True], ar[1:] != ar[:-1]))
-        return ar[flag]
+    def __ne__(self, other):
+        """Inequality comparison operator."""
+        return not self == other
+
+    def __lt__(self, other):
+        """Less-than comparison operator."""
+        if isinstance(other, ComparableArrayWrapper):
+            other = other.unwrapped
+        return self.unwrapped < other
+
+    def __gt__(self, other):
+        """Greather-than comparison operator."""
+        if isinstance(other, ComparableArrayWrapper):
+            other = other.unwrapped
+        return self.unwrapped > other
+
+    def __le__(self, other):
+        """Less-than-or-equal comparison operator."""
+        if isinstance(other, ComparableArrayWrapper):
+            other = other.unwrapped
+        return self.unwrapped <= other
+
+    def __ge__(self, other):
+        """Greater-than-or-equal comparison operator."""
+        if isinstance(other, ComparableArrayWrapper):
+            other = other.unwrapped
+        return self.unwrapped >= other
+
+    def __hash__(self):
+        """If the underlying object is hashable, the wrapper is too."""
+        return hash(self.unwrapped)
+
+    @staticmethod
+    def unwrap(v):
+        """Unwrap value if needed."""
+        return v.unwrapped if isinstance(v, ComparableArrayWrapper) else v
+
+
+def infer_dtype(values):
+    """Figure out dtype of sequence of sensor values, or None if unavailable."""
```

Comment:
I've improved the docstring.

There will basically always be a common dtype as determined by NumPy coercion, even if it is the fallback np.object. The dtype only remains unknown if there are no values to inspect.